cargo/workspace.bzl

```diff
@@ -0,0 +1,251 @@
+        repository_template = repository_ctx.attr.repository_template,
+        targets = ",\n".join(['"{}"'.format(target) for target in repository_ctx.attr.supported_targets]),
+        cargo = repository_ctx.path(cargo_label),
+    )
+
+    input_path = "_{name}.json".format(name = repository_ctx.attr.name)
+
+    # For debugging or working on changes to the resolver, you can set something like:
+    #   export RULES_RUST_RESOLVER_URL_OVERRIDE=file:///path/to/rules_rust/cargo/crate_universe_resolver/target/release/resolver
+    resolver_url = repository_ctx.os.environ.get("RULES_RUST_CRATE_UNIVERSE_RESOLVER_URL_OVERRIDE", None)
+    if resolver_url and resolver_url.startswith("file://"):
+        sha256_result = repository_ctx.execute(["sha256sum", resolver_url[7:]])
+        resolver_sha = sha256_result.stdout[:64]
+
+        resolver_path = repository_ctx.path("resolver")
+        repository_ctx.download(
+            url = resolver_url,
+            sha256 = resolver_sha,
+            output = resolver_path,
+            executable = True,
+        )
+    else:
+        # TODO: Avoid downloading both the linux and darwin one each time.
+        if repository_ctx.os.name == "linux":
+            resolver_path = repository_ctx.path(repository_ctx.attr._resolver_script_linux)
+        elif repository_ctx.os.name == "mac os x":
+            resolver_path = repository_ctx.path(repository_ctx.attr._resolver_script_darwin)
+
+    repository_ctx.file(input_path, content = input_content)
+
+    args = [
+        resolver_path,
+        "--input_path",
+        input_path,
+        "--output_path",
+        repository_ctx.path("defs.bzl"),
+        "--repo-name",
+        repository_ctx.attr.name,
+    ]
+    if lockfile_path != None:
+        args.append("--lockfile")
+        str(args.append(lockfile_path))
+    if repository_ctx.os.environ.get("RULES_RUST_UPDATE_CRATE_UNIVERSE_LOCKFILE", "false") == "true":
+        args.append("--update-lockfile")
+
+    path_binaries = [
+        repository_ctx.path(rustc_label),
+    ]
+
+    result = repository_ctx.execute(
+        args,
+        environment = {
+            "RUST_LOG": "info",
+
+            # The resolver invokes `cargo metadata` which relies on `rustc` being on the $PATH
+            "PATH": ":".join([str(b.dirname) for b in path_binaries]),
+        },
+    )
+    repository_ctx.delete(input_path)
+    if result.stderr:
+        print("Output from resolver: " + result.stderr)
+    if result.return_code != 0:
+        fail("Error resolving deps:\n" + result.stdout + "\n" + result.stderr)
+
+    repository_ctx.file("BUILD.bazel")
+
+_crate_universe_resolve = repository_rule(
+    implementation = _crate_universe_resolve_impl,
+    attrs = {
+        "packages": attr.string_list(allow_empty = True),
+        "cargo_toml_files": attr.label_list(),
+        "overrides": attr.string_dict(doc = """Mapping of crate name to crate.override(...) entries)
```

Comment:
This isn't going to render well when generated in docs, could instead a lot of the complex descriptions go in the top level `doc` attribute?

Additional nits:
- Begin docstrings like this with `"""\` and have content start on the next line
- Add `doc` field for all attributes
- Add `doc` field for rule